Internal modes

DavidSPhillips opened the pull request:
**Context:**
Calculating probability distributions and density matrices for partially distinguishable squeezed states with multiple internal modes through an interferometer with PNR detection.

Currently in `thewalrus` we only have the ability to simulate interference between Gaussian states which are either indistinguishable or fully distinguishable, and with only one internal mode per spatial mode. There exists no framework to simulate partially distinguishable states or states with more than one internal mode per spatial mode (e.g. when we have multiple Schmidt modes). There are code repos that already simulate internal modes and partial distinguishability, `two_squeezer_interference` for two degenerate squeezers, and `GKP_Multimode_Modelling` for the simulation of GKP states from GBS devices, however both those approaches use a combinatorial approach and therefore are inherently slow.

Using the same orthonormalisation procedures as those two named repos, the additions to this repo employs a new algorithm called the 'Internal Modes Haffnian' which cuts out unnecessary calculations from the combinatorial approach, and thus speeds up the calculations.

**Description of the Change:**
Nearly all of the changes are confined to the new directory `internal_modes` within the main `thewalrus` directory, and is reasonably self-contained and well documented. The unit tests for these functions live in the `test_internal_modes.py` script in the `tests` directory, and should be comprehensive though pending a codecov check. The main functions which will be the most user-facing are: `distinguishable_pnr_prob`, `density_matrix_single_mode`, and `pnr_prob`. The remainder of the functions are support functions.

The only change in the main `thewalrus` section is the Takagi decomposition. There is a function called `autonne` which lives in the `symplectic.py` script, which performs a Takagi-Autonne decomposition of a symmetric matrix. There is also a function in `strawberryfields` called `takagi` which performs the same decomposition using a different (and arguably more robust) algorithm. This function has been added to `thewalrus` with some additions to make it even more robust. Additionally, both functions now live in the `decompositions.py` script, and given they have different names they coexist without any problems. This decomposition is needed for some functions in the `internal_modes` directory, and it's the `takagi` version of the function that’s used there. Tests for both decomposition functions now live in `test_decompositions.py`.

**Benefits:**

- `thewalrus` now has the ability to simulate both multiple internal modes per spatial mode and partial distinguishability.

- Significantly faster computation for the listed specific system compared to the combinatorial approach in other repos.

- Ability to define the mode shape of the desired output mode (e.g. for homodyne detection).

- All dependencies on `strawberryfields` removed for the entire `thewalrus`.

**Possible Drawbacks:**

- No drawbacks to pre-existing code.

- New code only applies for squeezed states without displacement.

- For density matrix calculations we're restricted to post-selecting in one mode only (which is mandatory), and only a single internal mode is given (user can specify the basis function for this mode if they want).

- For photon-number distributions, only the probability for a particular pattern is given, the user must loop over all desired combinations for full distribution.

**Related GitHub Issues:**
None

jakeffbulmer committed:
first commit on internal_modes, with first draft of pnr_prob function

jakeffbulmer committed:
make numba import consistent

jakeffbulmer committed:
add fully distinguishable functionality

jakeffbulmer committed:
new name

jakeffbulmer committed:
include __init__ file

jakeffbulmer committed:
include single mode density matrix function

jakeffbulmer committed:
missing import

jakeffbulmer committed:
adding functionality for preparing internal mode covariance matrices

jakeffbulmer committed:
starting to add tests

jakeffbulmer committed:
new test comparing 3 methods for fully distinguishable GBS

jakeffbulmer committed:
adding new tests for distinguishable gbs

jakeffbulmer committed:
adding tests for density_matrix_single_mode with GKP circuits

jakeffbulmer committed:
heralded density matrix test

jakeffbulmer committed:
add some comments

jakeffbulmer committed:
not quite sure to be honest

jakeffbulmer committed:
fix prepare cov issues

DavidSPhillips committed:
Bug fix found from other repo

DavidSPhillips committed:
Black reformatting

DavidSPhillips committed:
Adding weights to matrix orthonormalisation

DavidSPhillips committed:
Making sure int32 arrays aren't causing problems in numba

DavidSPhillips committed:
Resolved merge conflict by using master version

DavidSPhillips committed:
Ignoring all pycache folders

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Can herald density matrix in arbitrary spatial mode

DavidSPhillips committed:
Option to add LO overlaps added to density matrix

DavidSPhillips committed:
Extra check added

DavidSPhillips committed:
Addition of hbar as a kwarg where applicable

DavidSPhillips committed:
Further changes to make density matrix compatible with numba jit

DavidSPhillips committed:
Removing pycache folder

DavidSPhillips committed:
Further tidying up

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Fixing issue with svd conditon

DavidSPhillips committed:
Moving LO overlap constraint check to non-jit function

DavidSPhillips committed:
removing pycache

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Updating internal mode test functions with updates

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Removing strawberryfields dependency on test_internal_modes

DavidSPhillips committed:
Fixing global phase on eigendecomposition and Takagi

DavidSPhillips committed:
Updating internal modes test functions

DavidSPhillips committed:
Fixing mistake in previous commits

DavidSPhillips committed:
Correcting decomposition issues and more real if close added

DavidSPhillips committed:
Takagi normalisation isn't going to work

DavidSPhillips committed:
Adding takagi decomposition similar to strawberryfields

DavidSPhillips committed:
Tidying up

DavidSPhillips committed:
Adding test for new Takagi function in symplectic

DavidSPhillips committed:
Adding autonne and takagi to autosummary

DavidSPhillips committed:
Tests for orthonormal_basis and state_prep added

DavidSPhillips committed:
Adding test function for prepare_cov

DavidSPhillips committed:
Readding sign convention to W matrices from Takagi output

DavidSPhillips committed:
Further improvements to Takagi function

DavidSPhillips committed:
Updates to docstring for Takagi function

DavidSPhillips committed:
Making location of test functions more obvious in test_internal_modes

DavidSPhillips committed:
Moving Autonne and Takagi from symplectic to decompositions

DavidSPhillips committed:
Missing comma added

DavidSPhillips committed:
Further missing commas added

DavidSPhillips committed:
Seeing if this solves the repoze.lru problem

DavidSPhillips committed:
Changing test internal modes takagi import to decompositions

DavidSPhillips committed:
Removing typing conventions

DavidSPhillips committed:
From itertools adding chain to test_internal_modes and groupby to decompositions

DavidSPhillips committed:
Fixing issue that causes NaNs in takagi

DavidSPhillips committed:
Fixing issue that causes NaNs in takagi in another place

DavidSPhillips committed:
Fixing takagi not reversing order when false and changing np.conjugates() to .conj()

Commit:
Passes black in all the files

Commit:
Making the linter a bit less unhappy

Commit:
Making the linter a bit less unhappy

Commit:
Making the linter a bit less unhappy

Commit:
Making the linter a bit less unhappy

Commit:
Passing black

nquesada commented on thewalrus/decompositions.py:
```diff
@@ -100,75 +320,3 @@ def williamson(V, rtol=1e-05, atol=1e-08):
-    if not is_symplectic(S):
-        raise ValueError("Input matrix is not symplectic.")
-
-    # Changing Basis
-    R = (1 / np.sqrt(2)) * np.block(
-        [[np.eye(N // 2), 1j * np.eye(N // 2)], [np.eye(N // 2), -1j * np.eye(N // 2)]]
-    )
-    Sc = R @ S @ np.conjugate(R).T
-    # Polar Decomposition
-    u1, d1, v1 = np.linalg.svd(Sc)
-    Sig = u1 @ np.diag(d1) @ np.conjugate(u1).T
-    Unitary = u1 @ v1
-    # Blocks of Unitary and Hermitian symplectics
-    alpha = Unitary[0 : N // 2, 0 : N // 2]
-    beta = Sig[0 : N // 2, N // 2 : N]
-    # Bloch-Messiah in this Basis
-    u2, d2, v2 = np.linalg.svd(beta)
-    sval = np.arcsinh(d2)
-    takagibeta = u2 @ sqrtm(np.conjugate(u2).T @ (v2.T))
-    uf = np.block([[takagibeta, 0 * takagibeta], [0 * takagibeta, np.conjugate(takagibeta)]])
-    vf = np.block(
-        [
-            [np.conjugate(takagibeta).T @ alpha, 0 * takagibeta],
-            [0 * takagibeta, np.conjugate(np.conjugate(takagibeta).T @ alpha)],
-        ]
-    )
-    df = np.block(
-        [
-            [np.diag(np.cosh(sval)), np.diag(np.sinh(sval))],
-            [np.diag(np.sinh(sval)), np.diag(np.cosh(sval))],
-        ]
-    )
-    # Rotating Back to Original Basis
-    uff = np.conjugate(R).T @ uf @ R
-    vff = np.conjugate(R).T @ vf @ R
-    dff = np.conjugate(R).T @ df @ R
-    dff = np.real_if_close(dff)
-    vff = np.real_if_close(vff)
-    uff = np.real_if_close(uff)
-    return uff, dff, vff
```
This should not be deleted

DavidSPhillips replied:
It hasn't been deleted, when I moved `autonne` here and added `takagi` I moved the functions such that they're in alphabetical order. All is still there, just reordered.

nquesada replied:
Was that necessary?

DavidSPhillips replied:
I saw that some other scripts had been orgnised in that way, so when I moved the functions I had the choice to just put them underneath or organise them in that way. I chose the latter.

Commit:
Making the linter a bit less unhappy

Commit:
Passing black

Commit:
Making pylint less unhappy

Commit:
Changes in _hafnian.py

Commit:
Changes atol

Commit:
Passes black

Commit:
simplfies code

Commit:
passes black

nquesada commented on thewalrus/internal_modes/useful_tools.py:
```diff
@@ -0,0 +1,140 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
```
maybe worth giving this module a better name.

sduquemesa replied:
the current one is too useful

nquesada commented on thewalrus/internal_modes/useful_tools.py:
```diff
@@ -0,0 +1,140 @@
+    return np.concatenate((xtmp1, xtmp2), axis=0)
+
+
+@numba.jit(nopython=True, cache=True)
+def f_all_charpoly(H, n):
+    """
+    Evaluate the polynomial coefficients of the function in the eigenvalue-trace formula.
+
+    Args:
+        H (array): matrix
+        n (int): number of polynomial coefficients to compute
+
+    Returns:
+        array: polynomial coefficients
+
+    """
+    pow_traces = powertrace(H, n // 2 + 1)
+    count = 0
+    comb = np.zeros((2, n // 2 + 1), dtype=np.complex128)
+    comb[0, 0] = 1
+    for i in range(1, n // 2 + 1):
+        factor = pow_traces[i] / (2 * i)
+        powfactor = 1
+        count = 1 - count
+        comb[count, :] = comb[1 - count, :]
+        for j in range(1, n // (2 * i) + 1):
+            powfactor *= factor / j
+            for k in range(i * j + 1, n // 2 + 2):
+                comb[count, k - 1] += comb[1 - count, k - i * j - 1] * powfactor
+
+    return comb[count, :]
```
this function already exists in `_hafnian.py`

rachelchadwick replied:
Yep but just called `f(A, n)` I think

nquesada commented on thewalrus/tests/test_internal_modes.py:
```diff
@@ -0,0 +1,1091 @@
+    sq_virt = params[9]
+
+    S1 = squeezing(np.abs(sq_r), phi=np.angle(sq_r))
+    U1 = np.array(
+        [
+            [np.cos(bs_theta1), -np.exp(-1j * bs_phi1) * np.sin(bs_theta1), 0],
+            [np.exp(1j * bs_phi1) * np.sin(bs_theta1), np.cos(bs_theta1), 0],
+            [0, 0, 1],
+        ]
+    )
+    U2 = np.array(
+        [
+            [1, 0, 0],
+            [0, np.cos(bs_theta2), -np.exp(-1j * bs_phi2) * np.sin(bs_theta2)],
+            [0, np.exp(1j * bs_phi2) * np.sin(bs_theta2), np.cos(bs_theta2)],
+        ]
+    )
+    U3 = np.array(
+        [
+            [np.cos(bs_theta3), -np.exp(-1j * bs_phi3) * np.sin(bs_theta3), 0],
+            [np.exp(1j * bs_phi3) * np.sin(bs_theta3), np.cos(bs_theta3), 0],
+            [0, 0, 1],
+        ]
+    )
+    Usymp = interferometer(U3 @ U2 @ U1)
+    r2 = np.array([0, 0, sq_virt])
+    S2 = squeezing(np.abs(r2), phi=np.angle(r2))
+    Z = S2 @ Usymp @ S1
+    cov = Z @ Z.T
```
you could do this using `expand` and `beamsplitter` from `thewalrus.symplectic`.

Better yet, we could omit all this and simply paste here the 6x6 covariance matrix representing the state.

DavidSPhillips replied:
Feel free to edit as you see fit :)

nquesada commented on thewalrus/tests/test_internal_modes.py:
```diff
@@ -0,0 +1,1091 @@
+    rho_loss2 = density_matrix_single_mode(cov_lossy, {1: m1, 2: m2}, cutoff=cutoff - 1)
+    rho_loss2 /= np.trace(rho_loss2)
+    print(np.max(np.abs(rho_loss1 - rho_loss2)))
+    assert np.allclose(rho_loss1, rho_loss2, atol=2.5e-4)
```
NOte that for this test to pass I had to put 2.5e-4 instead of the usual 1e-4.

Commit:
minor changes

Commit:
minor changes

Commit:
Most test pass. Had to remove memoization in guan code function

Commit:
Passes black

Commit:
Making the linter happy

DavidSPhillips committed:
Decreasing threshold for throwing vacuum modes away

DavidSPhillips committed:
Making black happy

DavidSPhillips committed:
Setting state prep thresh to 0 for tests

DavidSPhillips committed:
Black again

DavidSPhillips committed:
Setting prepare cov threshold to 0 for tests

DavidSPhillips committed:
Correcting error in previous commit

DavidSPhillips committed:
Fixing mistake in test density matrix

DavidSPhillips committed:
Changes to test density matrix

DavidSPhillips committed:
Changes to test density matrix again

DavidSPhillips committed:
Cutoff kwarg to test density matrix

DavidSPhillips committed:
Further changes to test density matrix

elib20 commented on thewalrus/tests/test_internal_modes.py:
```diff
@@ -0,0 +1,1098 @@
+    Z = S2 @ Usymp @ S1
+    cov = Z @ Z.T
+
+    cutoff = 26
```
I noticed that if the cutoff is changed to an odd number, the dimension of rho3 is rounded up to the next even number. Is that expected?

elib20 commented on thewalrus/tests/test_internal_modes.py:
```diff
@@ -0,0 +1,1098 @@
+    rho2 /= np.trace(rho2)
+
+    # get density matrix using new code
+    rho3 = density_matrix_single_mode(cov, {1: m1, 2: m2}, cutoff=cutoff - 1)
```
The Fock support of this output state should only be on even components. In rho1, the coefficients for the odd components are always 0. In rho2, the odd Fock support is at most order 1e-9. In rho3, some odd terms are as high as 1e-4, which seems to be the source of why @nquesada had to raise the tolerances to be so high.

DavidSPhillips committed:
Adding first test for Lowdin modes

DavidSPhillips committed:
Typo

DavidSPhillips committed:
Adding test for finding LO overlaps with Lowdin modes

DavidSPhillips committed:
Adding test for finding density matrix in specific LO mode

DavidSPhillips committed:
Adding extra import

DavidSPhillips committed:
Adding tqdm import

DavidSPhillips committed:
Removing tqdm entirely

DavidSPhillips committed:
Adding test for heradled single photon

rachelchadwick commented on thewalrus/internal_modes/distinguishable_pnr_prob.py:
```diff
@@ -0,0 +1,254 @@
+            yield K - i - 1, u[i], g[-2::-1]
+
+
+# @numba.jit(nopython=True, cache=True)
```
```suggestion
@numba.jit(nopython=True, cache=True)
```
Presumably we want this in?

DavidSPhillips replied:
@nquesada @sduquemesa was it removed for the codecov test?

nquesada replied:
I will look into this one!

rachelchadwick commented on thewalrus/internal_modes/distinguishable_pnr_prob.py:
```diff
@@ -0,0 +1,254 @@
+        # this could be done a tiny bit faster using gray code tricks
+        binom_prod = 1.0
+        for i in range(M):
+            binom_prod *= nb_binom(pattern[i], kept_edges[i])
```
```suggestion
            binom_prod *= nb_binom(np.int64(pattern[i]), np.int64(kept_edges[i]))
```
This fixes a numba problem for me (I think the commented out line 86). Alternatively the data type can be changed inside `nb_binom` which already exists or change `if k in (0, n)` to `if k==0 or k==n` in `nb_binom`. All options seem to work for me.

rachelchadwick commented on thewalrus/internal_modes/pnr_statistics.py:
```diff
@@ -0,0 +1,122 @@
+
+    M = As[0].shape[0] // (2 * K)
+
+    N = 2 * edge_reps.sum()
```
We could just define `N = edge_reps.sum()` as we mostly divide N by 2 afterwards anyway?

rachelchadwick commented on thewalrus/internal_modes/pnr_statistics.py:
```diff
@@ -0,0 +1,122 @@
+
+        binom_prod = 1.0
+        for i in range(M):
+            binom_prod *= nb_binom(edge_reps[i], kept_edges[i])
```
```suggestion
            binom_prod *= nb_binom(np.int64(edge_reps[i]), np.int64(kept_edges[i]))
```
Again for numba to work I need to change this or inside `nb_binom`.

rachelchadwick commented on thewalrus/internal_modes/pnr_statistics.py:
```diff
@@ -0,0 +1,122 @@
+        A[M * K :, :] = O[: M * K, :].conj()
+        As.append(A)
+
+        vac_prob /= np.sqrt(np.linalg.det(Q).real)
```
```suggestion
        vac_prob /= np.sqrt(np.real_if_close(np.linalg.det(Q)))
```

rachelchadwick commented on thewalrus/internal_modes/pnr_statistics.py:
```diff
@@ -0,0 +1,122 @@
+
+    fac_prod = np.prod(fac(i), dtype=np.float64)
+    haf = hafkd(np.array(As), np.array(i), K)
+    prob = haf.real * vac_prob / fac_prod
```
```suggestion
    prob = np.real_if_close(haf) * vac_prob / fac_prod
```

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    r"""The overlap matrix of all of the Schmidt modes in each spatial mode.
+
+    Args:
+        F (list[list/array[array]): List of arrays of the temporal modes of the Schmidt modes in each spatial mode, must be normalized and ordered by spatial mode.
```
The data type is missing a closing ']' somewhere. Should it be either list[list[array]] or list[array[array]] (rather that say list[list] or array[array])?

DavidSPhillips replied:
There's a missing ], it should be list[list/array[array]]

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    return O
+
+
+def vacuum_state(M, hbar=2):
+    r"""
+    Returns the displacement vector and covariance matrix for an M mode vacuum state.
+
+    Args:
+        M (int): number of modes (spatial, Schmidt, other etc.)
+        hbar (float): (default 2) the value of :math:`\hbar` in the commutation relation :math:`[\x,\p]=i\hbar`
+
+    Returns:
+        tuple[array]: the displacement vector and covariance matrix of an M mode vacuum state
+    """
+
+    return np.zeros(2 * M), np.eye(2 * M) * hbar / 2
```
Why not just import `vacuum_state` from `symplectic`?

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+
+
+# pylint: disable=too-many-branches
+def orthonormal_basis(rjs, O=None, F=None, thr=1e-3):
```
If both O and F are given (ie not `None`) it looks like F is used and O is ignored. Should we add a warning that O is ignored or a check that they are compatible (ie O=O_matrix(F)) if both are given?

DavidSPhillips replied:
Sure we can add that

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    rs = np.array([r for rj in rjs for r in rj])
+    if F is not None:
+        Fs = [f for fj in F for f in fj]
+        Fs = [Fs[j] / np.linalg.norm(Fs[j]) for j in range(len(Fs))]
```
Do we need to divide by the norm if F is normalised (as in the docstring description)?

DavidSPhillips replied:
It's me being overcautious...

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    if F is not None:
+        Fs = [f for fj in F for f in fj]
+        Fs = [Fs[j] / np.linalg.norm(Fs[j]) for j in range(len(Fs))]
+        if not np.allclose(len(Fs), rs.shape[0]):
```
```suggestion
        if len(Fs) != rs.shape[0]:
```
Why not just compare the two integers exactly?

DavidSPhillips replied:
It would be absolutely fine in this case, but I have an aversion to using anything other than np.allclose for comparisons in general

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    elif O is not None:
+        if not np.allclose(O, O.conj().T):
+            raise ValueError("O must be a Hermitian matrix")
+        if not np.allclose(O.shape[0], rs.shape[0]):
```
```suggestion
        if O.shape[0] != rs.shape[0]:
```

DavidSPhillips replied:
Same here

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+        raise ValueError("Either F or O must be given")
+    R = O.shape[0]
+    M = len(rjs)
+    njs = np.array([len(rjs[i]) for i in range(M)])
```
```suggestion
    njs = np.array([len(i) for i in rjs])
```

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    """
+    if not np.allclose(len(eps), len(W)):
+        raise ValueError("len of eps must be equal to len of W")
+    epsBig = np.array([])
+    for i in eps:
+        epsBig = np.append(epsBig, i)
```
```suggestion
    epsBig = np.concatenate(eps, axis=0)
```

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+    M = len(eps)
+    R = eps[0].shape[0]
+
+    Qvac = (hbar / 2) * np.eye(2 * M * R)
+    S = squeezing(epsBig)
+    Qinit = S @ Qvac @ S.T
+
+    Wbig = np.eye(M * R, dtype=np.complex128)
+    for j in range(M):
+        Wbig[R * j : R * (j + 1), R * j : R * (j + 1)] = W[j]
+
+    Q = interferometer(Wbig.T.conj()) @ Qinit @ interferometer(Wbig.T.conj()).T
+    Qswap = interferometer(swap_matrix(M, R)) @ Q @ interferometer(swap_matrix(M, R)).T
```
All the Q matrices here are not the same as the Q matrix from `Qmat` to my knowledge. The notation might be a bit confusing.

DavidSPhillips replied:
The other option is to call them all cov (or similar), think it's worth the effort?

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+
+    Q = interferometer(Wbig.T.conj()) @ Qinit @ interferometer(Wbig.T.conj()).T
+    Qswap = interferometer(swap_matrix(M, R)) @ Q @ interferometer(swap_matrix(M, R)).T
+    muVac, covVac = vacuum_state(M, hbar=hbar)
```
We could put this line after `if R > 1:` as it isn't used otherwise.

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+
+def LO_overlaps(chis, LO_shape):
+    r"""
+    Computes the overlap integral between the
```
This docstring is a bit suspenseful...

rachelchadwick commented on thewalrus/internal_modes/prepare_cov.py:
```diff
@@ -0,0 +1,304 @@
+        [
+            np.inner(
+                LO_shape.conj() / np.linalg.norm(LO_shape),
+                chis[j] / np.linalg.norm(chis[j]),
+            )
+            for j in range(len(chis))
```
```suggestion
                j / np.linalg.norm(j),
            )
            for j in chis
```

DavidSPhillips replied:
Implemented something similar

rachelchadwick commented on thewalrus/internal_modes/useful_tools.py:
```diff
@@ -0,0 +1,140 @@
+import numba
+from ..charpoly import powertrace
+
+fact = np.array([factorial(i) for i in range(110)], dtype=np.float64)
```
Why find the factorials up to 110? Wherever this is called, could we not choose the upper limit based on the detection pattern and cutoff (ie only calculate what is needed)?

rachelchadwick commented on thewalrus/internal_modes/useful_tools.py:
```diff
@@ -0,0 +1,140 @@
+    then schmidt_reps=[1,1,1,2,2,2]
+
+    Args:
+        spatial_reps (array): number of spatial reps
```
What is reps?

DavidSPhillips replied:
Representation. All the modes are ordered by default to internal modes then spatial modes, e.g. all the internal modes in the first spatial mode, then all the internal modes in the second spatial mode, and so on. I think this function just expands from just spatial modes to all modes in the default order.

rachelchadwick commented on thewalrus/internal_modes/useful_tools.py:
```diff
@@ -0,0 +1,140 @@
+        array: reps of schmidt modes
+    """
+
+    M = len(spatial_reps)
+    schmidt_reps = np.empty(M * K, dtype=spatial_reps.dtype)
+    for i, r in enumerate(spatial_reps):
+        schmidt_reps[i * K : (i + 1) * K] = r
+
+    return schmidt_reps
```
```suggestion
    return np.repeat(spatial_reps, K)
```

DavidSPhillips replied:
Not sure if this would affect the @numba.jit(nopython=True, cache=True) above the function or not, anyone know?

rachelchadwick commented on thewalrus/tests/test_internal_modes.py:
```diff
@@ -0,0 +1,1369 @@
+        * np.exp(-1j * phi)
+        / np.sqrt(2)
+    )
+    eps, W = [np.array([r, 0]), np.array([r, 0])], [W0, W1]
```
This line can be deleted I think.

DavidSPhillips committed:
Some extra checks

DavidSPhillips committed:
Update thewalrus/internal_modes/prepare_cov.py

Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

DavidSPhillips committed:
Update thewalrus/internal_modes/prepare_cov.py

Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

DavidSPhillips committed:
Changing Q to cov for readability

DavidSPhillips committed:
Further changes for clarity and reducing unnecessary calculations

DavidSPhillips committed:
Using inbuilt function for  0), pack-reused 0

DavidSPhillips committed:
More changes for clarity and readability

DavidSPhillips committed:
Removing TODO as it's already done

DavidSPhillips committed:
Fix in nb_binom for windows

DavidSPhillips committed:
Chaning the name of useful_tools to utils

DavidSPhillips committed:
Making black happy

DavidSPhillips committed:
Using beam_splitter and expand instead of writing explicit unitaries in test_internal_modes

DavidSPhillips committed:
Testing O matrix explicitly and adding more coverage for testing

DavidSPhillips committed:
Improving pylint

DavidSPhillips committed:
More things for pylint

DavidSPhillips committed:
More pylint disabling for things that won't be changed

sduquemesa commented on thewalrus/internal_modes/distinguishable_pnr_prob.py:
```diff
@@ -0,0 +1,254 @@
+@numba.jit(nopython=True, cache=True)
+def f_all(E, n):
+    """
+    evaluate the polyonial coefficients of the function in the eigevalue-trace formula
```
```suggestion
    evaluate the polynomial coefficients of the function in the eigenvalue trace formula
```

sduquemesa commented on thewalrus/internal_modes/distinguishable_pnr_prob.py:
```diff
@@ -0,0 +1,254 @@
+    """
+    evaluate the polyonial coefficients of the function in the eigevalue-trace formula
+    Args:
+        E (array): eigevavlues of AX
```
```suggestion
        E (array): eigenvalues of AX
```

nquesada committed:
Update thewalrus/internal_modes/distinguishable_pnr_prob.py

Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

rachelchadwick committed:
Internal modes update (#355)

* density_matrix_single_mode returns expected dimension

* density_matrix_single_mode shouldn't include cutoff

* Remove repeated f functions

* Making black happy

* Remove distinguishable_pnr_prob

* Run black

* Simplify tests

* Run black

* Only reordering eigenvalues and vectors when all eigenvalues are not identical

* Changing way we rearrange eigenvalues in descending order

* Added prob check for different cov method

* Restore pragma: no cover comment

* Increase tolerances

Co-authored-by: David Phillips <62555251+DavidSPhillips@users.noreply.github.com>

DavidSPhillips committed:
Adding assertion tests to density_matrix_internal_modes

DavidSPhillips committed:
black

DavidSPhillips committed:
Adding assertion tests to orthonormal_basis

DavidSPhillips committed:
Adding assertion tests to state_prep

DavidSPhillips committed:
Adding assertion tests to prepare_cov

DavidSPhillips committed:
Further codecov stuff

DavidSPhillips committed:
Further codecov stuff again

DavidSPhillips committed:
Setting normalize=False as default in internal modes desnity matrix

DavidSPhillips committed:
Running black through selected files

DavidSPhillips committed:
Normalising density matrices for tests

DavidSPhillips committed:
Extra normalize added to test_mixed_heralded_photon

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Seeing if this solves black issue

DavidSPhillips committed:
More black issues

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Merge branch 'master' into internal_modes

DavidSPhillips committed:
Merge branch 'master' into internal_modes

nquesada committed:
Merging master into internal_modes (#365)

* Refactor sphinx config and bump docs python version (#364)

* fix [.readthedocs.yml] - increase python version

Fixes urllib3 v2.0 not supporting versions <1.1.1 of openSSL. See https://github.com/readthedocs/readthedocs.org/issues/10290.

* fix [.readthedocs.yml] - set container image and python version

* unpin doc requirements

* Update decompositions.py and readds the Takagi decomposition (#363)

* Update decompositions.py

* New takagi

* New takagi

* Passes black

* Simplifies blochmessiah

* Simplifies blochmessiah

* Found a case that breaks Takagi

* Fixes all the tests

* Fixes issues found by the linter

* Adds extra test

* Adds extra test

* dummy

* dummy

* fix [.readthedocs.yml] - increase python version

Fixes urllib3 v2.0 not supporting versions <1.1.1 of openSSL. See https://github.com/readthedocs/readthedocs.org/issues/10290.

* fix [.readthedocs.yml] - set container image and python version

* unpin doc requirements

* Revert "fix [.readthedocs.yml] - increase python version"

This reverts commit db0e8d7c6fc6838e0aa2126b3ae813fc2397fcba.

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

* Dummy commit

* Passes black

* linting

* Updates module summary

---------

Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>
Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

nquesada committed:
Merge branch 'master' into internal_modes

nquesada committed:
Merge branch 'master' into internal_modes

nquesada committed:
Merge branch 'master' into internal_modes

nquesada committed:
Adding probability tests (#368)

* First working version still lacks wrapper, tests

* Test pass

* Adds more tests

* Adds more tests

* Adds more tests

* Add clarification of docstring

* Adds doscstring

* Apply suggestions from code review

Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

* passes black

* Apply suggestions from code review

* Add relative import

* Increase tolerance

* Increase tolerance

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

nquesada committed:
Merge branch 'master' into internal_modes

elib20 committed:
Merge branch 'master' into internal_modes

nquesada committed:
Off diagonal (#383)

* First commit

* Fixes bug

* removes unused import

* Working version

* Passes black

* removes unused function

* removes unused function

* Rachels code review (#385)

* Minor fixes

* passes black

* Fix test

* removes unused function

* test error is raised

* test error is raised

* Passes black

* Passes black

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

* Now the non-recursive function do not produce warning (#386)

* Now the non-recursive function do not produce warning

* Now the non-recursive function do not produce warning

* Passes black

* Simplifies the diagonal method

* More tests, always

* More tests, always

* minor pylint improvements

* Adds extra test

* black

* minor simplification

* some tests pass

* some tests pass

* Moves the tests that take forever to the end of the test file

* Passes black

* Adds else

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

* Adds the warning (#388)

* Adds the warning

* passes black

* Apply suggestions from code review

Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

* Update thewalrus/internal_modes/fock_density_matrices.py

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

* Windows fix for numba types (#389)

* Fix numba types

* Passes black

* Apply suggestions from code review

* Apply suggestions from code review

* passes black

* Make lists tuples

* Check non-recursive

* Tests with high tolerances

* Run black

* Update test_density_matrix

* Minor changes

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
Co-authored-by: Nicolas Quesada <991946+nquesada@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

* removes unnecesary import

* Squashed commit of the following:

commit 7c63905c9623684237091dc5a2779fabea1e79e9
Merge: 127a1a7 f935053
Author: Eli Bourassa <53090166+elib20@users.noreply.github.com>
Date:   Mon Apr 29 14:16:22 2024 -0400

    Merge branch 'master' into internal_modes

commit f935053ad9065ad505781298bcd0734534705ab9
Author: Alberto Fumagalli <alberto@xanadu.ai>
Date:   Mon Apr 29 14:14:51 2024 -0400

    Add codecov token to test workflow (#390)

commit fe9e1124ca902050408e426f42b642ebb2930982
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Wed Apr 10 15:31:34 2024 -0400

    Updates references (#384)

    * updates references

    * Updates biblio

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

commit 25b5f0820520865a311de2e7d44779e7338adc1d
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Wed Feb 28 20:35:28 2024 -0500

    Implements the pre-Iwasawa and Iwasawa decompositions. (#382)

    * First working version

    * First working version

    * First working version

    * better tests

    * better tests

    * Finally correct

    * Finally correct

    * putting the traingles in the right places

    * Removes unnecesary T

    * More tests

    * More tests and spell correction

    * More tests and spell correction

    * Removes unnecesary complex number usage

    * Apply suggestions from code review

    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

    * Apply suggestions from code review

    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

commit 127a1a77c91710c52254f52efdb571a1869fa361
Merge: e6de063 8759363
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Tue Feb 13 16:43:20 2024 -0500

    Merge branch 'master' into internal_modes

commit 875936316fb2e34b9a4e9e778c8a7b9a705367a3
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Thu Feb 1 16:57:23 2024 -0500

    Better blochmessiah (#381)

    * Nicer BM

    * Updates changelog

    * Unnecesary import

    * More simplifications

    * More simplifications

    * one more

    * Updates docstring

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

commit 6247fc885d3902a7a5aab7e7dbb63687614b1c09
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Wed Jan 24 10:43:57 2024 -0500

    Updates docstring and simplifies Williamson (#380)

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

commit 0b1af633a0d745ebb896c4f01c6499873ff16bb0
Author: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>
Date:   Thu Jan 11 17:09:32 2024 -0500

    Increment version number to `0.22.0-dev` (#379)

commit 544c4570905d165b5f3194017ea42bd8b10f424e
Author: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>
Date:   Wed Jan 10 16:52:38 2024 -0500

    bump version to `0.21.0` (#378)

commit 2268bf04c8e1abf4c9089a63cddd4d334eb9cf5c
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Thu Jan 4 11:12:27 2024 -0500

    Adds extra degenerate tests with nonvac nullspace (#377)

    * Adds extra degenerate tests with nonvac nullspace

    * updates changelog

    * updates changelog

    * updates changelog

    * Removes unnecesary comment

    * Still breaking something

    * Simplifications still work

    * Simplifications still work

    * Fixes bug

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

commit 8ec2172e5bdb48a7d32886742fe3e7c829e48fd9
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Wed Nov 29 15:53:11 2023 -0500

    Implements the Montrealer (#374)

    * Saving changes

    * Passes black

    * Passes black

    * Black

    * updates acknowledgements and changelog

    * updates acknowledgements and changelog

    * Adds montrealer

    * Adds Montrealer tests (#375)

    * Yanic the pirate

    * mtl test

    * mtl first tests

    * mtl test

    * test mtl

    * test montrealer

    * test montrealer

    * test Montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * tests montrealer

    * montrealer test

    * montrealer test

    * montrealer test

    * montrealer test

    * montrealer test

    * test montrealer

    * test montrealer

    * montrealer test

    * montrealer test

    * montrealer test

    * montrealer tests

    * montrealer tests

    * montrealer tests

    * montrealer tests

    * functions header update

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer ready for review

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer - failed tests

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * test montrealer

    * the rspms are now ordered

    * Passes black

    * garbage test file to be deleted later

    * minor changes

    * Done

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

    * making bots happy

    * Apply suggestions from code review

    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

    * Apply suggestions from code review

    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

    * Apply suggestions from code review

    * Apply suggestions from code review

    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

    * Apply suggestions from code review

    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

    * Adds extra tests

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
    Co-authored-by: Yanic Cardin <yanic.cardin.1@gmail.com>
    Co-authored-by: Sebastián Duque Mesa <675763+sduquemesa@users.noreply.github.com>

commit 0e163bfb3c4bb92b691d35b2e25ea817719132e4
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Tue Aug 22 10:19:08 2023 -0400

    Update symplectic.py (#372)

    * Update decompositions.py

    * New takagi

    * New takagi

    * Passes black

    * Simplifies blochmessiah

    * Simplifies blochmessiah

    * Found a case that breaks Takagi

    * Fixes all the tests

    * Fixes issues found by the linter

    * Adds extra test

    * Adds extra test

    * dummy

    * dummy

    * Update symplectic.py

    Removes `autonne`

    * relocates test

    * trying to make pylint happy

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

commit e999e4937c75977b5bee7ae33fbf5afb718aec35
Author: Nicolas Quesada <991946+nquesada@users.noreply.github.com>
Date:   Thu Aug 17 12:55:45 2023 -0400

    Better way to determine if a matrix is a phase times a real matrix (#373)

    * Add a better way to determine if a matrix is a phase times a real matrix and tests

    * CHANGELOG updates

    ---------

    Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>

* Revert "Squashed commit of the following:"

This reverts commit 0f99b20f90473f30d42a30c4851dd6653dc5ce67.

---------

Co-authored-by: Nicolas Quesada <nquesada@pop-os.localdomain>
Co-authored-by: rachelchadwick <33873960+rachelchadwick@users.noreply.github.com>

timmysilv committed:
fix segfaults

timmysilv commented:
the cause of the segfault has been found! You cannot [use asserts](https://github.com/XanaduAI/thewalrus/blob/c390f84cbb1fee76a8d1a71225efbddcc14c173c/thewalrus/internal_modes/fock_density_matrices.py#L106) within `numba.prange` loop contexts as described in the second note of [their docs on parallelization](https://numba.readthedocs.io/en/latest/user/parallel.html#explicit-parallel-loops). Once I removed that line, I was told: `Use of reduction variable 'haf_arr' other than in a supported reduction function is not permitted` - updated the loop to not use it, and everything worked!

nquesada commented:
Wow. That is crazy! Thanks so much!

timmysilv committed:
add assert back as comment for posterity

RyosukeNORO committed:
Merge branch 'master' into internal_modes

nquesada commented on thewalrus/internal_modes/fock_density_matrices.py:
```diff
@@ -0,0 +1,277 @@
+            for i in range(cutoff):
+                for j in range(i + 1):
+                    if (i - j) % 2 == 0:
+                        patt_long = (j,) + N_nums + ((i - j) // 2,)
```
@ajmartinezc

nquesada commented:
Hi Luke --- From memory, there is an issue with which numba version to use. Perhaps @rachelchadwick remembershow to proceed and which version to use. Otherwise, I will be immensely happy to see this being merged 3+ years.
